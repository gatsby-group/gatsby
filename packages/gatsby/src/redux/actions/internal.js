/**
 * @typedef {Object} Plugin
 * @param {string} name
 */

const actions = {}

/**
 * Create a dependency between a page and data. Probably for
 * internal use only.
 * @param {Object} $0
 * @param {string} $0.path the path to the page
 * @param {string} $0.nodeId A node ID
 * @param {string} $0.connection A connection type
 * @param {string} plugin
 * @private
 */
actions.createPageDependency = ({ path, nodeId, connection }, plugin = ``) => {
  return {
    type: `CREATE_COMPONENT_DEPENDENCY`,
    plugin,
    payload: {
      path,
      nodeId,
      connection,
    },
  }
}

/**
 * Delete dependencies between an array of pages and data. Probably for
 * internal use only. Used when deleting pages.
 * @param {string[]} paths the paths to delete.
 * @private
 */
actions.deleteComponentsDependencies = paths => {
  return {
    type: `DELETE_COMPONENTS_DEPENDENCIES`,
    payload: {
      paths,
    },
  }
}

/**
 * When the query watcher extracts a GraphQL query, it calls
 * this to store the query with its component.
 * @private
 *
 * @param {Object} $0
 * @param {string} $0.query
 * @param {string} $0.componentPath
 */
actions.replaceComponentQuery = ({ query, componentPath }) => {
  return {
    type: `REPLACE_COMPONENT_QUERY`,
    payload: {
      query,
      componentPath,
    },
  }
}

/**
 * When the query watcher extracts a "static" GraphQL query from <StaticQuery>
 * components, it calls this to store the query with its component.
 * @private
 *
 * @param {*} args
 * @param {Plugin | null} [plugin]
 */
actions.replaceStaticQuery = (args, plugin = null) => {
  return {
    type: `REPLACE_STATIC_QUERY`,
    plugin,
    payload: args,
  }
}

/**
 *
 * Report that a query has been extracted from a component. Used by
 * query-compiler.js.
 *
 * @param {Object} $0
 * @param {componentPath} $0.componentPath The path to the component that just had
 * its query read.
 * @param {query} $0.query The GraphQL query that was extracted from the component.
 * @param {Plugin} plugin
 * @param {string} [traceId]
 * @private
 */
actions.queryExtracted = ({ componentPath, query }, plugin, traceId) => {
  return {
    type: `QUERY_EXTRACTED`,
    plugin,
    traceId,
    payload: { componentPath, query },
  }
}

/**
 *
 * Report that the Relay Compiler found a graphql error when attempting to extract a query
 *
 * @param {Object} $0
 * @param {componentPath} $0.componentPath The path to the component that just had
 * its query read.
 * @param {error} $0.error The GraphQL query that was extracted from the component.
 * @param {Plugin} plugin
 * @param {string} [traceId]
 * @private
 */
actions.queryExtractionGraphQLError = (
  { componentPath, error },
  plugin,
  traceId
) => {
  return {
    type: `QUERY_EXTRACTION_GRAPHQL_ERROR`,
    plugin,
    traceId,
    payload: { componentPath, error },
  }
}

/**
 *
 * Report that babel was able to extract the graphql query.
 * Indicates that the file is free of JS errors.
 *
 * @param {Object} $0
 * @param {componentPath} $0.componentPath The path to the component that just had
 * its query read.
 * @param {Plugin} plugin
 * @param {string} [traceId]
 * @private
 */
actions.queryExtractedBabelSuccess = ({ componentPath }, plugin, traceId) => {
  return {
    type: `QUERY_EXTRACTION_BABEL_SUCCESS`,
    plugin,
    traceId,
    payload: { componentPath },
  }
}

/**
 *
 * Report that the Relay Compiler found a babel error when attempting to extract a query
 *
 * @param {Object} $0
 * @param {componentPath} $0.componentPath The path to the component that just had
 * its query read.
 * @param {error} $0.error The Babel error object
 * @param {Plugin} plugin
 * @param {string} [traceId]
 * @private
 */
actions.queryExtractionBabelError = (
  { componentPath, error },
  plugin,
  traceId
) => {
  return {
    type: `QUERY_EXTRACTION_BABEL_ERROR`,
    plugin,
    traceId,
    payload: { componentPath, error },
  }
}

/**
 * Set overall program status e.g. `BOOTSTRAPING` or `BOOTSTRAP_FINISHED`.
 *
 * @param {string} status - Program status
 * @param {Plugin} plugin
 * @param {string} [traceId]
 * @private
 */
actions.setProgramStatus = (status, plugin, traceId) => {
  return {
    type: `SET_PROGRAM_STATUS`,
    plugin,
    traceId,
    payload: status,
  }
}

/**
 * Broadcast that a page's query was run.
 *
 * @param {Object} $0
 * @param {componentPath} $0.componentPath The path to the component that just had
 * its query read.
 * @param {string} $0.path to the page component that changed.
 * @param {boolean} $0.isPage
 * @param {Plugin} plugin
 * @param {string} [traceId]
 * @private
 */
actions.pageQueryRun = ({ path, componentPath, isPage }, plugin, traceId) => {
  return {
    type: `PAGE_QUERY_RUN`,
    plugin,
    traceId,
    payload: { path, componentPath, isPage },
  }
}

module.exports = { actions }
