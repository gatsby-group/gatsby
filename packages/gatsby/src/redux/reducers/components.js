const _ = require(`lodash`)
const normalize = require(`normalize-path`)

module.exports = (state = new Map(), action) => {
  switch (action.type) {
    case `DELETE_CACHE`:
      return new Map()
    case `CREATE_PAGE`:
      action.payload.componentPath = normalize(action.payload.component)
      console.log(
        `CREATE_PAGE`,
        action.payload.componentPath,
        action.payload.query
      )
      state.set(
        action.payload.componentPath,
        _.merge(
          { query: ``, queryState: `QUERY_NOT_YET_EXTRACTED` },
          state.get(action.payload.componentPath),
          {
            componentPath: action.payload.componentPath,
          }
        )
      )
      return state
    case `REMOVE_TEMPLATE_COMPONENT`:
      state.delete(normalize(action.payload.componentPath))
      return state
    case `REPLACE_COMPONENT_QUERY`:
      console.log(
        `REPLACE_COMPONENT_QUERY`,
        action.payload.componentPath,
        action.payload.query
      )
      action.payload.componentPath = normalize(action.payload.componentPath)
      state.set(action.payload.componentPath, {
        queryState: `QUERY_EXTRACTED`,
        ...state.get(action.payload.componentPath),
        query: action.payload.query,
      })
      return state
  }

  return state
}
