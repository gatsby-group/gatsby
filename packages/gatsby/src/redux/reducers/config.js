const chalk = require(`chalk`)

const { didYouMean } = require(`../../utils/did-you-mean`)
const { gatsbyConfigSchema } = require(`../../joi-schemas/joi`)

module.exports = (state = {}, action) => {
  switch (action.type) {
    case `SET_SITE_CONFIG`: {
      // Validate the config.
      const result = gatsbyConfigSchema.validate(action.payload || {})

      const normalizedPayload = result.value

      // TODO use Redux for capturing errors from different
      // parts of Gatsby so a) can capture richer errors and b) be
      // more flexible how to display them.
      if (result.error) {
        console.log(
          chalk.blue.bgYellow(`The site's gatsby-config.js failed validation`)
        )

        const [invalidProperty] = result.error.details.filter(
          details => details.type === `object.allowUnknown`
        )

        if (invalidProperty) {
          const { key } = invalidProperty.context
          const suggestion = didYouMean(key)

          throw new Error(`${result.error}. ${suggestion}`)
        } else {
          console.log(chalk.bold.red(result.error))
        }

        if (normalizedPayload.linkPrefix) {
          console.log(`"linkPrefix" should be changed to "pathPrefix"`)
        }
        throw new Error(`The site's gatsby-config.js failed validation`)
      }

      return normalizedPayload
    }
    default:
      return state
  }
}
