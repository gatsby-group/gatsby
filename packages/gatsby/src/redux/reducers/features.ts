import { IToggleFeatureAction, IGatsbyState } from "../types"

export const initialState: IGatsbyState["features"] = {
  imageService: false,
}

export const featuresReducer = (
  state: IGatsbyState["features"] = Object.assign({}, initialState),
  action: IToggleFeatureAction
): IGatsbyState["features"] => {
  switch (action.type) {
    case `TOGGLE_FEATURE`: {
      // when a feature got enabled we don't want plugins to disable it
      if (initialState[action.payload.name] !== action.payload.value) {
        state[action.payload.name] = action.payload.value
      }

      return state
    }
    default:
      return state
  }
}
