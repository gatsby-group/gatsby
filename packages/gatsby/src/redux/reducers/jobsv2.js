module.exports = (
  state = { active: new Set(), stale: new Map(), done: new Map() },
  action
) => {
  switch (action.type) {
    case `CREATE_JOB_V2`: {
      const { job, plugin } = action.payload

      state.stale.set(job.contentDigest, {
        job,
        plugin,
      })
      state.active.add(job.contentDigest)

      return state
    }

    case `END_JOB_V2`: {
      const { job, result } = action.payload
      state.stale.delete(job.contentDigest)
      state.active.delete(job.contentDigest)
      // inputPaths is used to make sure the job is still necessary
      state.done.set(job.contentDigest, { result, inputPaths: job.inputPaths })

      return state
    }

    case `REMOVE_STALE_JOB_V2`: {
      const { contentDigest } = action.payload
      state.stale.delete(contentDigest)
      state.done.delete(contentDigest)

      return state
    }
  }

  return state
}
