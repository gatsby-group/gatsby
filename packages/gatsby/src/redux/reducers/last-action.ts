import { IGatsbyState, ActionsUnion } from "../types"

export const lastActionReducer = (
  _state: IGatsbyState["lastAction"],
  action: ActionsUnion
): IGatsbyState["lastAction"] => action
