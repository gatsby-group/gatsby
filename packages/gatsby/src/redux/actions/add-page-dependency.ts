import { store } from "../"

import { createPageDependency as internalCreatePageDependency } from "./internal"

export const createPageDependency = ({
  path,
  nodeId,
  connection,
}: Parameters<typeof internalCreatePageDependency>[0]): void => {
  const {
    componentDataDependencies: { connections, nodes },
  } = store.getState()

  // Check that the dependencies aren't already recorded so we
  // can avoid creating lots of very noisy actions.
  const nodeDependencyExists =
    !nodeId || (nodeId && nodes.has(nodeId) && nodes.get(nodeId).has(path))
  const connectionDependencyExists =
    !connection ||
    (connection &&
      connections.has(connection) &&
      connections.get(connection).has(path))

  if (nodeDependencyExists && connectionDependencyExists) {
    return
  }

  // It's new, let's dispatch it
  store.dispatch(internalCreatePageDependency({ path, nodeId, connection }))
}
