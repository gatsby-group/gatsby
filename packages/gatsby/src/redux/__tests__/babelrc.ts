import { actions } from "../actions"
import { babelrcReducer } from "../reducers/babelrc"
import {
  prepareOptions,
  mergeConfigItemOptions,
  convertCustomPresetsToPlugins,
} from "../../utils/babel-loader-helpers"

describe(`Babelrc actions/reducer`, () => {
  it(`allows adding a new plugin`, () => {
    const action = actions.setBabelPlugin(
      { name: `test-babel-plugin` },
      { name: `test` }
    )
    expect(babelrcReducer(undefined, action)).toMatchSnapshot()
  })

  it(`allows updating the options of an existing plugin`, () => {
    const action = actions.setBabelPlugin(
      { name: `test-babel-plugin`, options: { id: 1 } },
      { name: `test` }
    )
    let state = babelrcReducer(undefined, action)
    expect(state.stages.develop.plugins[0].options.id).toBe(1)

    const updateAction = actions.setBabelPlugin(
      { name: `test-babel-plugin`, options: { id: 2 } },
      { name: `test` }
    )
    state = babelrcReducer(state, updateAction)
    expect(state.stages.develop.plugins[0].options.id).toBe(2)
  })

  it(`allows adding a new preset`, () => {
    const action = actions.setBabelPreset(
      { name: `test-babel-preset` },
      { name: `test` }
    )
    expect(babelrcReducer(undefined, action)).toMatchSnapshot()
  })

  it(`allows updating the options of an existing preset`, () => {
    const action = actions.setBabelPreset(
      { name: `test-babel-preset`, options: { id: 1 } },
      { name: `test` }
    )
    let state = babelrcReducer(undefined, action)
    expect(state.stages.develop.presets[0].options.id).toBe(1)

    const updateAction = actions.setBabelPreset(
      { name: `test-babel-preset`, options: { id: 2 } },
      { name: `test` }
    )
    state = babelrcReducer(state, updateAction)
    expect(state.stages.develop.presets[0].options.id).toBe(2)
  })

  it(`allows specifying the stage for the plugin`, () => {
    const action = actions.setBabelPlugin(
      { name: `test-babel-plugin`, stage: `build-javascript` },
      { name: `test` }
    )
    const state = babelrcReducer(undefined, action)
    expect(state.stages.develop.plugins.length).toBe(0)
    expect(state.stages[`build-javascript`].plugins.length).toBe(1)
  })

  it(`allows specifying the stage for the preset`, () => {
    const action = actions.setBabelPreset(
      { name: `test-babel-preset`, stage: `build-javascript` },
      { name: `test` }
    )
    const state = babelrcReducer(undefined, action)
    expect(state.stages.develop.presets.length).toBe(0)
    expect(state.stages[`build-javascript`].presets.length).toBe(1)
  })

  it(`sets default presets/plugins if there's no userland babelrc`, () => {
    const fakeResolver = (moduleName): string => `/path/to/module/${moduleName}`
    const babel = { createConfigItem: jest.fn() }

    prepareOptions(babel, { stage: `test` }, fakeResolver)

    expect(babel.createConfigItem.mock.calls).toMatchSnapshot()
  })

  it(`adds stage option to babel-preset-gatsby defined with userland babelrc`, () => {
    const fakeResolver = (moduleName): string => `/path/to/module/${moduleName}`

    const BabelPresetGatsbyConfigItem = {
      file: { resolved: fakeResolver(`babel-preset-gatsby`) },
    }

    const babel = {
      createConfigItem: jest.fn((presetDescriptor: any): any => {
        if (!Array.isArray(presetDescriptor)) {
          presetDescriptor = [presetDescriptor]
        }

        const name = presetDescriptor[0]
        if (name.includes(`babel-preset-gatsby`)) {
          return {
            ...BabelPresetGatsbyConfigItem,
            options: presetDescriptor[1],
          }
        }
        return undefined
      }),
    }
    const presets: any = [BabelPresetGatsbyConfigItem]

    const { presets: convertedPresets } = convertCustomPresetsToPlugins(
      babel,
      { presets, plugins: [] },
      {
        stage: `develop`,
        reactRuntime: `automatic`,
        reactImportSource: `@emotion/react`,
      },
      fakeResolver
    )

    const babelPresetGatsbyPreset = convertedPresets.find(preset =>
      preset?.file?.resolved?.includes(`babel-preset-gatsby`)
    )

    expect(babelPresetGatsbyPreset?.options).toMatchInlineSnapshot(`
      Object {
        "reactImportSource": "@emotion/react",
        "reactRuntime": "automatic",
        "stage": "develop",
      }
    `)
  })

  it(`adds stage option to babel-preset-gatsby defined in nested preset`, () => {
    const fakeResolver = (moduleName): string => `/path/to/module/${moduleName}`

    const BabelPresetGatsbyConfigItem = {
      file: { resolved: fakeResolver(`babel-preset-gatsby`) },
    }

    const CustomPresetGatsbyConfigItem = {
      file: { resolved: fakeResolver(`custome-preset`) },
      value: (): any => {
        return {
          presets: [`babel-preset-gatsby`],
          plugins: [[`babel-plugin-testing-stuff`, { foo: `bar` }]],
        }
      },
    }

    const babel = {
      createConfigItem: jest.fn((presetDescriptor: any): any => {
        if (!Array.isArray(presetDescriptor)) {
          presetDescriptor = [presetDescriptor]
        }

        const name = presetDescriptor[0]
        if (name.includes(`babel-preset-gatsby`)) {
          return {
            ...BabelPresetGatsbyConfigItem,
            options: presetDescriptor[1],
          }
        }

        if (name.includes(`babel-plugin-testing-stuff`)) {
          return {
            options: presetDescriptor[1],
            file: { resolved: fakeResolver(`babel-plugin-testing-stuff`) },
          }
        }

        return undefined
      }),
    }
    const presets: any = [CustomPresetGatsbyConfigItem]

    const { presets: convertedPresets, plugins: convertedPlugins } =
      convertCustomPresetsToPlugins(
        babel,
        { presets, plugins: [] },
        {
          stage: `develop`,
          reactRuntime: `automatic`,
          reactImportSource: `@emotion/react`,
        },
        fakeResolver
      )

    const babelPresetGatsbyPreset = convertedPresets.find(preset =>
      preset?.file?.resolved?.includes(`babel-preset-gatsby`)
    )

    expect(babelPresetGatsbyPreset?.options).toMatchInlineSnapshot(`
      Object {
        "reactImportSource": "@emotion/react",
        "reactRuntime": "automatic",
        "stage": "develop",
      }
    `)

    // make sure we preserve options for plugins from the custom preset
    const pluginFromCustomPreset = convertedPlugins.find(plugin =>
      plugin?.file?.resolved?.includes(`babel-plugin-testing-stuff`)
    )
    expect(pluginFromCustomPreset?.options).toMatchInlineSnapshot(`
      Object {
        "foo": "bar",
      }
    `)
  })

  it(`prefers reactRuntime/reactImportSource from preset options over ones from gatsby-config`, () => {
    const fakeResolver = (moduleName): string => `/path/to/module/${moduleName}`

    const BabelPresetGatsbyConfigItem = {
      file: { resolved: fakeResolver(`babel-preset-gatsby`) },
      options: {
        reactRuntime: `classic`,
      },
    }

    const babel = {
      createConfigItem: jest.fn((presetDescriptor: any): any => {
        if (!Array.isArray(presetDescriptor)) {
          presetDescriptor = [presetDescriptor]
        }

        const name = presetDescriptor[0]
        if (name.includes(`babel-preset-gatsby`)) {
          return {
            ...BabelPresetGatsbyConfigItem,
            options: presetDescriptor[1],
          }
        }
        return undefined
      }),
    }
    const presets: any = [BabelPresetGatsbyConfigItem]

    const { presets: convertedPresets } = convertCustomPresetsToPlugins(
      babel,
      { presets, plugins: [] },
      {
        stage: `develop`,
        reactRuntime: `automatic`,
        reactImportSource: `@emotion/react`,
      },
      fakeResolver
    )

    const babelPresetGatsbyPreset = convertedPresets.find(preset =>
      preset?.file?.resolved?.includes(`babel-preset-gatsby`)
    )

    expect(babelPresetGatsbyPreset?.options).toMatchInlineSnapshot(`
      Object {
        "reactRuntime": "classic",
        "stage": "develop",
      }
    `)
  })

  it(`allows setting options`, () => {
    const action = actions.setBabelOptions(
      { options: { sourceMaps: `inline` } },
      { name: `test` }
    )
    let state = babelrcReducer(undefined, action)
    expect(state.stages.develop.options.sourceMaps).toBe(`inline`)

    const updateAction = actions.setBabelOptions(
      { options: { sourceMaps: true } },
      { name: `test` }
    )
    state = babelrcReducer(state, updateAction)

    expect(state.stages.develop.options.sourceMaps).toBe(true)
  })

  it(`allows setting options on a particular stage`, () => {
    const action = actions.setBabelOptions(
      { options: { sourceMaps: `inline` }, stage: `develop` },
      { name: `test` }
    )
    const state = babelrcReducer(undefined, action)
    expect(state.stages.develop.options.sourceMaps).toBe(`inline`)
    expect(state.stages[`develop-html`].options.sourceMaps).toBe(undefined)
  })

  it(`allows merging config items`, () => {
    const babel = { createConfigItem: jest.fn() }
    // This merges in new change.
    mergeConfigItemOptions({
      items: [{ options: { wat: 1 }, file: { resolved: `hi` } }],
      itemToMerge: { options: { wat: 2 }, file: { resolved: `hi` } },
      type: `plugin`,
      babel,
    })
    expect(babel.createConfigItem.mock.calls).toMatchSnapshot()

    expect(
      mergeConfigItemOptions({
        items: [{ options: { wat: 1 }, file: { resolved: `hi` } }],
        itemToMerge: { options: { wat: 2 }, file: { resolved: `hi2` } },
        type: `plugin`,
        babel,
      })
    ).toMatchSnapshot()
  })
})
