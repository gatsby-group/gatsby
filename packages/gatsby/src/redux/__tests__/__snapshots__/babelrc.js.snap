// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Babelrc actions/reducer allows adding a new plugin 1`] = `
Object {
  "stages": Object {
    "build-html": Object {
      "options": Object {
        "babelrc": false,
        "cacheDirectory": true,
        "sourceType": "unambiguous",
      },
      "plugins": Array [
        Object {
          "name": "test-babel-plugin",
          "options": Object {},
        },
      ],
      "presets": Array [],
    },
    "build-javascript": Object {
      "options": Object {
        "babelrc": false,
        "cacheDirectory": true,
        "sourceType": "unambiguous",
      },
      "plugins": Array [
        Object {
          "name": "test-babel-plugin",
          "options": Object {},
        },
      ],
      "presets": Array [],
    },
    "develop": Object {
      "options": Object {
        "babelrc": false,
        "cacheDirectory": true,
        "sourceType": "unambiguous",
      },
      "plugins": Array [
        Object {
          "name": "test-babel-plugin",
          "options": Object {},
        },
      ],
      "presets": Array [],
    },
    "develop-html": Object {
      "options": Object {
        "babelrc": false,
        "cacheDirectory": true,
        "sourceType": "unambiguous",
      },
      "plugins": Array [
        Object {
          "name": "test-babel-plugin",
          "options": Object {},
        },
      ],
      "presets": Array [],
    },
  },
}
`;

exports[`Babelrc actions/reducer allows adding a new preset 1`] = `
Object {
  "stages": Object {
    "build-html": Object {
      "options": Object {
        "babelrc": false,
        "cacheDirectory": true,
        "sourceType": "unambiguous",
      },
      "plugins": Array [],
      "presets": Array [
        Object {
          "name": "test-babel-preset",
          "options": Object {},
        },
      ],
    },
    "build-javascript": Object {
      "options": Object {
        "babelrc": false,
        "cacheDirectory": true,
        "sourceType": "unambiguous",
      },
      "plugins": Array [],
      "presets": Array [
        Object {
          "name": "test-babel-preset",
          "options": Object {},
        },
      ],
    },
    "develop": Object {
      "options": Object {
        "babelrc": false,
        "cacheDirectory": true,
        "sourceType": "unambiguous",
      },
      "plugins": Array [],
      "presets": Array [
        Object {
          "name": "test-babel-preset",
          "options": Object {},
        },
      ],
    },
    "develop-html": Object {
      "options": Object {
        "babelrc": false,
        "cacheDirectory": true,
        "sourceType": "unambiguous",
      },
      "plugins": Array [],
      "presets": Array [
        Object {
          "name": "test-babel-preset",
          "options": Object {},
        },
      ],
    },
  },
}
`;

exports[`Babelrc actions/reducer allows merging config items 1`] = `
Array [
  Array [
    Array [
      "hi",
      Object {
        "wat": 2,
      },
    ],
    Object {
      "type": "plugin",
    },
  ],
]
`;

exports[`Babelrc actions/reducer allows merging config items 2`] = `
Array [
  Object {
    "file": Object {
      "resolved": "hi",
    },
    "options": Object {
      "wat": 1,
    },
  },
  Object {
    "file": Object {
      "resolved": "hi2",
    },
    "options": Object {
      "wat": 2,
    },
  },
]
`;

exports[`Babelrc actions/reducer sets default presets/plugins if there's no userland babelrc 1`] = `
Array [
  Array [
    Array [
      "/Users/kylemathews/programs/gatsby/packages/babel-plugin-remove-graphql-queries/index.js",
    ],
    Object {
      "type": "plugin",
    },
  ],
  Array [
    Array [
      "/Users/kylemathews/programs/gatsby/node_modules/@babel/preset-env/lib/index.js",
      Object {
        "loose": true,
        "modules": false,
        "targets": Object {
          "browsers": undefined,
        },
        "useBuiltIns": "usage",
      },
    ],
    Object {
      "type": "preset",
    },
  ],
  Array [
    Array [
      "/Users/kylemathews/programs/gatsby/node_modules/@babel/preset-react/lib/index.js",
      Object {
        "development": false,
        "pragma": "React.createElement",
        "useBuiltIns": true,
      },
    ],
    Object {
      "type": "preset",
    },
  ],
  Array [
    Array [
      "/Users/kylemathews/programs/gatsby/node_modules/@babel/preset-flow/lib/index.js",
    ],
    Object {
      "type": "preset",
    },
  ],
  Array [
    Array [
      "/Users/kylemathews/programs/gatsby/node_modules/@babel/plugin-proposal-class-properties/lib/index.js",
      Object {
        "loose": true,
      },
    ],
    Object {
      "type": "plugin",
    },
  ],
  Array [
    Array [
      "/Users/kylemathews/programs/gatsby/node_modules/babel-plugin-macros/dist/index.js",
    ],
    Object {
      "type": "plugin",
    },
  ],
  Array [
    Array [
      "/Users/kylemathews/programs/gatsby/node_modules/@babel/plugin-syntax-dynamic-import/lib/index.js",
    ],
    Object {
      "type": "plugin",
    },
  ],
  Array [
    Array [
      "/Users/kylemathews/programs/gatsby/node_modules/@babel/plugin-transform-runtime/lib/index.js",
      Object {
        "helpers": true,
        "polyfill": false,
        "regenerator": true,
      },
    ],
    Object {
      "type": "plugin",
    },
  ],
]
`;
